### PR TITLE
Enable shellcheck to run for select scripts

### PR DESCRIPTION
**Pull request by dinatale2**
### Description
Enable shellcheck to run on zed scripts, paxcheck.sh, zfs-tests.sh, zfs.sh, and zloop.sh.

### Motivation and Context

We would like to perform shellcheck on scripts contained within the ZFS repository. This will help enforce good style and help avoid simple bugs in shell scripts now and in the future. We also plan to enable shellcheck on the ZFS buildbot in the future.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: cmd/zed/zed.d/zed-functions.sh [hunk 1]
```diff
@@ -29,8 +30,8 @@ ZED_FLOCK_FD=8
 #
 zed_check_cmd()
 {
-    local cmd
-    local rv=0
+    cmd=$*
+    rv=0
```
Comment (behlendorf) [inline cmd/zed/zed.d/zed-functions.sh]: `dash`, `bash`, and `ash` all support the local keyword.  So even though this isn't strictly POSIX sh it's OK for us to leave this as is.  Let' not make any of these local keyword changes.  I should add that leaving these as local doesn't generate any shellcheck errors for me with shellcheck-0.3.7.  If you're getting warnings let's exclude them with `shellcheck -e`.

File: scripts/paxcheck.sh [hunk 2]
```diff
@@ -1,14 +1,16 @@
 #!/bin/sh
 
-if ! type scanelf > /dev/null 2>&1; then
+SCANELF="/usr/bin/scanelf"
```
Comment (behlendorf) [inline scripts/paxcheck.sh]: We can't safely assume paths.  Plus `dash`, `ash`, and `bash` all support type.  Let's disable the warning.

File: scripts/paxcheck.sh [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
+if ! [ -x "$SCANELF" ] > /dev/null 2>&1; then
     echo "scanelf (from pax-utils) is required for these checks." >&2
     exit 3
 fi
 
 RET=0
 
 # check for exec stacks
-OUT="$(scanelf -qyRAF '%e %p' $1)"
+OUT=$($SCANELF -qyRAF '%e %p' "$1")
```
Comment (behlendorf) [inline scripts/paxcheck.sh]: No need for the s/scanelf/SCANELF change here

File: scripts/zfs.sh [hunk 4]
```diff
@@ -44,12 +46,14 @@ while getopts 'hvud' OPTION; do
...
 		;;
 	v)
+		# shellcheck disable=SC2034
 		VERBOSE=1
 		;;
 	u)
 		UNLOAD=1
 		;;
 	d)
+		# shellcheck disable=SC2034
```
Comment (behlendorf) [inline scripts/zfs.sh]: This really is unused dead functionality.  The spl logging facility has been removed, this and the spl_dump_log() function should go.

File: scripts/zloop.sh [hunk 5]
```diff
@@ -56,9 +58,8 @@ function usage
 
 function or_die
 {
-	$@
-	if [[ $? -ne 0 ]]; then
-		echo "Command failed: $@"
+	if ! "$*"; then
+		echo "Command failed: $*"
```
Comment (behlendorf) [inline scripts/zloop.sh]: This is intentional.  Disable the warning.

File: scripts/zloop.sh [hunk 6]
```diff
@@ -220,14 +222,14 @@ while [[ $timeout -eq 0 ]] || [[ $curtime -le $(($starttime + $timeout)) ]]; do
 	zopt="$zopt -s $size"
 	zopt="$zopt -f $workdir"
 
-	cmd="$ZTEST $zopt $@"
+	cmd="$ZTEST $zopt $*"
```
Comment (behlendorf) [inline scripts/zloop.sh]: Same let's disable the warning.

**Comment by tonyhutter** [comment]
@dinatale2 I'll give this a test on our enclosure

File: cmd/zed/zed.d/all-debug.sh [hunk 7]
```diff
@@ -5,7 +5,9 @@
 # This can be a useful aid when developing/debugging ZEDLETs since it shows the
 # environment variables defined for each zevent.
 
+#shellcheck source=/dev/null
```
Comment (dinatale2) [inline cmd/zed/zed.d/all-debug.sh]: `warning: Can't follow non-constant source. Use a directive to specify location. [SC1090]` is the error that is produced without them.

Comment (behlendorf) [inline cmd/zed/zed.d/all-debug.sh]: Let's just exclude these errors for now by adding `--exclude=SC1090` to the command line.

File: cmd/zed/zed.d/zed-functions.sh (hunk 1 above)
Comment (dinatale2) [inline cmd/zed/zed.d/zed-functions.sh]: These errors came from adding `#!/bin/sh` as the first line of the file. `warning: In POSIX sh, 'local' is undefined. [SC2039]` is the error that results for the local keyword. I could turn the warning off for the file, but `SC2039` is the generic undefined warning, so we'd be losing that coverage.

Comment (behlendorf) [inline cmd/zed/zed.d/zed-functions.sh]: That's unfortunate, but we can live with that for the moment.

**Comment by dinatale2** [comment]
@tonyhutter In case you've started testing this, I just refreshed this, so you may want to grab the new changes.

**Comment by dinatale2** [comment]
I went ahead and tested my changes in our testbed. I'm not seeing changes in behavior and the output seems to be very close as compared to what is in master. @tonyhutter would you mind testing this again? I'm still not all that confident that I didn't break anything.

**Comment by tonyhutter** [comment]
@dinatale2 I was able to get it to work again with the changes below.  I dunno if my changes break shellcheck syntax though:
```
diff --git a/cmd/zed/zed.d/statechange-led.sh b/cmd/zed/zed.d/statechange-led.sh
index 56e4e72..32333d2 100755
--- a/cmd/zed/zed.d/statechange-led.sh
+++ b/cmd/zed/zed.d/statechange-led.sh
@@ -24,6 +24,7 @@
 #   2: enclosure leds administratively disabled
 #   3: The led sysfs path passed from ZFS does not exist
 #   4: $ZPOOL not set
+#   5: awk not installed
 
 [ -f "${ZED_ZEDLET_DIR}/zed.rc" ] && . "${ZED_ZEDLET_DIR}/zed.rc"
 . "${ZED_ZEDLET_DIR}/zed-functions.sh"
@@ -37,6 +38,7 @@ if [ "${ZED_USE_ENCLOSURE_LEDS}" != "1" ] ; then
 fi
 
 zed_check_cmd "$ZPOOL" || exit 4
+zed_check_cmd awk || exit 5
 
 # Global used in set_led debug print
 vdev=""
@@ -52,7 +54,7 @@ vdev=""
 # Return
 #  0 on success, 3 on missing sysfs path
 #
-check_and_set_led
+check_and_set_led ()
 {
        file="$1"
        val="$2"
@@ -86,7 +88,7 @@ check_and_set_led
         done
 }
 
-state_to_val
+state_to_val ()
 {
        state="$1"
        if [ "$state" = "FAULTED" ] || [ "$state" = "DEGRADED" ] || \
@@ -108,7 +110,7 @@ state_to_val
 # Return
 #  0 on success, 3 on missing sysfs path
 #
-process_pool
+process_pool ()
 {
        pool="$1"
        rc=0
@@ -116,15 +118,31 @@ process_pool
        # Lookup all the current LED values and paths in parallel
        #shellcheck disable=SC2016
        cmd='echo led_token=$(cat "$VDEV_ENC_SYSFS_PATH/fault"),"$VDEV_ENC_SYSFS_PATH",'
-       out=$($ZPOOL status -vc "$cmd" "$pool" | grep 'led_token=')
+
+       if [ -z "$pool" ] ; then
+               out=$($ZPOOL status -vc "$cmd" | grep 'led_token=')
+       else
+               zed_log_msg "running on pool $pool"
+               out=$($ZPOOL status -vc "$cmd" "$pool" | grep 'led_token=')
+       fi
+
        #shellcheck disable=SC2034
-       while read -r vdev state read write chksum therest; do
+       echo "$out" | while read -r vdev state read write chksum therest; do
                current_val=
                vdev_enc_sysfs_path=
 
                # Read out current LED value and path
                tmp=$(echo "$therest" | sed 's/^.*led_token=//g')
-               IFS="," read -r current_val vdev_enc_sysfs_path < echo "$tmp"
+               current_val=$(echo "$tmp" | awk -F ',' '{print $1}')
+
+               # On some enclosures if you write 1 to fault, and read it back,
+               # it will return 2.  Treat all non-zero values as 1 for
+               # simplicity.
+               if [ "$current_val" != "0" ] ; then
+                       current_val=1
+               fi
+
+               vdev_enc_sysfs_path=$(echo "$tmp" | awk -F ',' '{print $2}')
 
                if [ -z "$vdev_enc_sysfs_path" ] ; then
                        # Skip anything with no sysfs LED entries
@@ -137,7 +155,7 @@ process_pool
                        continue;
                fi
 
-               val=$(state_to_val "$state")
+               val="$(state_to_val "$state")"
 
                if [ "$current_val" = "$val" ] ; then
                        # LED is already set correctly
@@ -148,9 +166,7 @@ process_pool
                        rc=1
                fi
 
-       done << EOF
-$out
-EOF
+       done 
 
        if [ "$rc" = "0" ] ; then
                return 0
@@ -168,5 +184,5 @@ if [ ! -z "$ZEVENT_VDEV_ENC_SYSFS_PATH" ] && [ ! -z "$ZEVENT_VDEV_STATE_STR" ] ;
 else
        # Process the entire pool
        poolname=$(zed_guid_to_pool "$ZEVENT_POOL_GUID")
-       process_pool "$poolname"
+       process_pool $poolname
 fi
diff --git a/cmd/zed/zed.d/zed-functions.sh b/cmd/zed/zed.d/zed-functions.sh
index 8480ae8..b7de510 100644
--- a/cmd/zed/zed.d/zed-functions.sh
+++ b/cmd/zed/zed.d/zed-functions.sh
@@ -427,7 +427,7 @@ zed_rate_limit()
 # Return
 #   Pool name
 #
-zed_guid_to_pool
+zed_guid_to_pool()
 {
        if [ -z "$1" ] ; then
                return
```

**Comment by dinatale2** [comment]
@tonyhutter thanks for taking a look! Just refreshed my patch. The zed scripts should now function correctly and shellcheck should now pass if it were enabled.

**Pull request merged**